### PR TITLE
parser: move scan.go to new scanner package 

### LINKED ISSUE #64710: sql/parser: split `scan.go` into a separate package with few dependencies
(Discussed with @petermattis)

We wish to embed the SQL tokenizer (not parser) in a 3rd party go code base for the CC management console, with minimal dependencies on the remainder of the CockroachDB code base.

For this we want to move the scanner to a new package with minimal go imports: `pkg/sql/scanner`.

The code in `pkg/sql/parser/scan.go` is already pretty well self-contained.

What remains to be done is to **remove the imports for`pkg/sql/lex` and `pkg/sql/sem/tree`** and **stop relying on pkg/sql/parser/sql.go**.

before we can create a new `pkg/sql/scanner`.

How to achieve this?

## Regarding the `pkg/sql/lex` dependency [done]

The only use of it is for `lex.GetKeywordID()`. This function is a "leaf" function and does not need to remain in the `lex` package. It can be moved to `lexbase` or a similar self-contained package.

## Regarding the `pkg/sql/sem/tree` dependency [done]

There are just two uses: `tree.NewNumVal` and `tree.NewPlaceholder`.

We cannot really move these functions to a different package to make them standalone, because the `tree.NumVal` and `tree.NewPlaceholder` are (currently) intimately linked to other things in `sem/tree`.

What we can do however is **dependency injection**: define two global function callbacks in the new `pkg/sql/scanner` package:

```go
var NewNumValFn = func(constant.Value, string, bool) interface{} { return &fakeNum{}  }
var NewPlaceholderFn = func(string) interface{} { return &fakePlaceholder{} }
```

(where the `fakeNum` and `fakePlaceholder` types are suitable defined locally in `scanner`)

Then in the `sql/parser` package, define an `init` function:

```go
func init() {
   scanner.NewNumValFn = func(a constant.Value, s string, b bool) interface{} { return tree.NewNumVal(a,s,b) }
   scanner.NewPlaceholderFn = func(s string) interface{} { return tree.NewPlaceholder(s) }
}
```

## Regarding `pkg/sql/parser/sql.go`

This is a generated file based on sql.y. scan.go uses it to get the `sqlSymType` and `sqlSymUnion` types. pkg/sql/parser/sql.go has a lot of dependencies. We need to extract `sqlSymType` and `sqlSymUnion` to a separate package.

Epic CRDB-7217

### PR DESCRIPTION
**Pull request by rafiss**
Fixes https://github.com/cockroachdb/cockroach/issues/64710

See individual commits.

This extracts an interface for sqlSymType so that scan.go can keep using it.

This allows scan.go to live in its own package so that other components can
use its functionality without needing to pull in a large dependency
on sql/sem/tree.

Release note: None

### CODE REVIEW

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -152,6 +152,41 @@ func processUnaryQualOpInternal(
...
 
 %{
+// sqlSymType is generated by goyacc, and implements the ScanSymType interface.
+var _ ScanSymType = &sqlSymType{}
+
+func (s *sqlSymType) Id() int32 {
+	return s.id
+}
+
+func (s *sqlSymType) SetId(id int32) {
```
Comment (otan) [inline pkg/sql/parser/sql.y:163]: should we call this `SetID`?

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -152,6 +152,41 @@ func processUnaryQualOpInternal(
 %}
 
 %{
+// sqlSymType is generated by goyacc, and implements the ScanSymType interface.
+var _ ScanSymType = &sqlSymType{}
+
+func (s *sqlSymType) Id() int32 {
```
Comment (otan) [inline pkg/sql/parser/sql.y:159]: `ID`?

**Review by otan — COMMENTED** [review]
lgtm mod interface function names

**Review by knz — COMMENTED** [review]
Change LGTM but I'd like to try to trick git (and reviewable) into recognizing the file rename, so that we don't get a completely fresh new file without any history.

I have checked locally and it appears that you can achieve that by renaming `parser/scan.go` into `parser/scanner.go` and amending the 2nd commit. Does that work for you too?

> Reviewed 3 of 3 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67584#-:-MeZeGvKDnEcIwqNQu4W:b6tppvk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @rafiss)

**Comment by rafiss** [comment]
> I have checked locally and it appears that you can achieve that by renaming parser/scan.go into parser/scanner.go and amending the 2nd commit. Does that work for you too?

that worked! thanks

**Review by knz — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r=knz,otan

**Pull request merged**